### DISCUSSION TITLE
Unrecognized field name "compNames" with ravenCobraWrapper error

### DISCUSSION THREAD

**Discussion by PkiwiBird** [SysBioChalmers/RAVEN#521] [Help me!] [answered]
Hi,
when trying to convert a raven model to a cobra model with ravenCobraWrapper, I get the following error:
```
Converting COBRA structure to RAVEN..
Unrecognized field name "compNames".

Error in ravenCobraWrapper (line 216)
        newModel.mets=regexprep(newModel.mets,['\[', model.compNames{i}, '\]$'],'');
```
Do you have any idea on how to solve this?

**Comment by edkerk** [comment]
How did you load this RAVEN model (with what command)? And would you be able to link to its source (where did you get the model from)?

**Reply by PkiwiBird** [reply]
It was solved by using _compNames_ and _comps_ as fields but now there's another problem.

The model is a liver model extracted from the complete body model. Inside _[Dataset EV1](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7285886/bin/MSB-16-e8982-s022.zip)_ there is the male whole body model: _MSB-16-e8982-s022/PSCM_toolbox/SuppliedResults/OrganAtlas_Harvey.mat_.
The following code is used to extract the liver model and save as .yml: 
```
load 'OrganAtlas_Harvey.mat'
omc_ml=OrganCompendium_male; % organ model compendium of male
liver_ml=omc_ml.Liver.modelAllComp; % liver model of male
compNames = {'cytosol'; 'lysosome'; 'mitochondria'; 'endoplasmatic reticulum';
'golgi apparatus'; 'peroxisome'; 'nucleus'; 'inner mitochondria space';
'bile duct liver'; 'blood portal vein liver'; 'blood portal vein';
'blood circulation'; 'bile duct'};
comps = {'c'; 'l'; 'm'; 'r'; 'g'; 'x'; 'n'; 'i'; 'bdL'; 'bpL'; 'bp'; 'bc'; 'bd'};
liver_ml.compNames = compNames;
liver_ml.comps = comps;
liver_ml = ravenCobraWrapper(liver_ml);
writeYAMLmodel(liver_ml, fullfile(BaseDir,'results', 'liver_ml_GEM.yml'));
```
The last command gives the error:
```
Brace indexing is not supported for variables of this type.

Error in writeYAMLmodel>writeField (line 276)
                    list{j} = ['"' list{j} '"'];

Error in writeYAMLmodel (line 100)
    writeField(model, fid, 'subSystems',           'txt', i, '  - subsystem',             preserveQuotes)

```
Could you help point in the right direction, please?

**Reply by edkerk** [reply, marked as answer]
First, in your original post you mention
> convert a RAVEN model to a COBRA model

but the model is already in COBRA format.

But if you meant to write "COBRA to RAVEN model", then:

1. Do not define the `comps` and `compNames` fields as done above, you'd want `ravenCobraWrapper` to define them and appropriately refer to them in `metComps`.
2. After `ravenCobraWrapper` is finished, manually corrected the entries in `liver_ml.compNames`, as they will by default be the same abbreviations as in `.comps`.
3. To deal with the above and the odd mixture of nested cell arrays that are used for the subSystems field, I've made some changes to `ravenCobraWrapper` and `readYAMLmodel`.

So first get these two functions from PR #522, and then run:

```matlab
load 'OrganAtlas_Harvey.mat'
omc_ml=OrganCompendium_male; % organ model compendium of male
liver_ml=omc_ml.Liver.modelAllComp; % liver model of male
liver_ml = ravenCobraWrapper(liver_ml);
%% Manually curate the liver_ml.compNames field before continuing
writeYAMLmodel(liver_ml, fullfile(BaseDir,'results', 'liver_ml_GEM.yml'));
```

**Reply by edkerk** [reply]
This is now also turned into issue #523

**Reply by PkiwiBird** [reply]
Thank you very much!
